GTFS RT validation:  Investigate `[Errno 2] No such file or directory` issues

`[Errno 2] No such file or directory` is our most common cause of validation failures. (In the midnight UTC hour on 6/29/23, occurred for 13 distinct URLs, 4 of them for the entire hour.)
See for example:
```
SELECT
  name,
  SUBSTR(validation_exception, 0, 35),
  COUNT(*)
FROM `cal-itp-data-infra.staging.stg_gtfs_rt__vehicle_positions_validation_outcomes` 
WHERE dt = '2023-06-29' AND hour = '2023-06-29T00:00:00' AND process_stderr IS NULL AND NOT validation_success
GROUP BY 1,2
ORDER BY 3 DESC
```
The goal of this ticket is to try to investigate why this is happening.

Per @atvaccaro:
> IIRC these are situations where the validator is not writing an output JSON file to the expected path

When running in the multithreaded configuration, two classes of errors are visible:
1. `[Errno2] No such file or directory` (686 occurrences vs 130 when not multithreaded)
2. The following decompression failure, which does not occur when running single-threaded (another variant is visible when 

```
[main] INFO org.mobilitydata.gtfsrtvalidator.lib.batch.BatchProcessor - Reading GTFS data from /var/folders/_l/q3hq8y3s17xc6vl_wjk9tgy00000gn/T/tmpanj1jr_x/google_transit.zip...
Exception in thread "main" org.onebusaway.csv_entities.exceptions.CsvEntityIOException: io error: entityType=org.onebusaway.gtfs.model.StopTime path=stop_times.txt lineNumber=360508
	at org.onebusaway.csv_entities.CsvEntityReader.readEntities(CsvEntityReader.java:161)
	at org.onebusaway.csv_entities.CsvEntityReader.readEntities(CsvEntityReader.java:120)
	at org.onebusaway.csv_entities.CsvEntityReader.readEntities(CsvEntityReader.java:115)
	at org.onebusaway.gtfs.serialization.GtfsReader.run(GtfsReader.java:171)
	at org.onebusaway.gtfs.serialization.GtfsReader.run(GtfsReader.java:159)
	at org.mobilitydata.gtfsrtvalidator.lib.batch.BatchProcessor.readGtfsData(BatchProcessor.java:315)
	at org.mobilitydata.gtfsrtvalidator.lib.batch.BatchProcessor.processFeeds(BatchProcessor.java:139)
	at org.mobilitydata.gtfsrtvalidator.lib.Main.main(Main.java:62)
Caused by: java.io.EOFException: Unexpected end of ZLIB input stream
	at java.base/java.util.zip.ZipFile$ZipFileInflaterInputStream.fill(ZipFile.java:447)
	at java.base/java.util.zip.InflaterInputStream.read(InflaterInputStream.java:159)
	at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:333)
	at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:376)
	at java.base/sun.nio.cs.StreamDecoder.lockedRead(StreamDecoder.java:219)
	at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:173)
	at java.base/java.io.InputStreamReader.read(InputStreamReader.java:189)
	at java.base/java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.base/java.io.BufferedReader.implReadLine(BufferedReader.java:371)
	at java.base/java.io.BufferedReader.readLine(BufferedReader.java:348)
	at java.base/java.io.BufferedReader.readLine(BufferedReader.java:437)
	at org.onebusaway.csv_entities.CsvEntityReader.readEntities(CsvEntityReader.java:148)
	... 7 more
```

--- a/jobs/gtfs-rt-parser-v2/gtfs_rt_parser.py
+++ b/jobs/gtfs-rt-parser-v2/gtfs_rt_parser.py
@@ -61,6 +61,18 @@ class InvalidMetadata(Exception):
     pass
 
 
+class NoScheduleDataSpecified(Exception):
+    pass
+
+
+class ScheduleDataNotFound(Exception):
+    pass
+
+
+class NoValidatorResults(Exception):
+    pass
+
+
 class RTProcessingStep(str, Enum):
     parse = "parse"
     validate = "validate"
@@ -75,14 +87,6 @@ class RTValidationMetadata(BaseModel):
     extract_ts: pendulum.DateTime
     extract_config: GTFSDownloadConfig
     gtfs_validator_version: str
-
-
-class NoScheduleDataSpecified(Exception):
-    pass
-
-
-class ScheduleDataNotFound(Exception):
-    pass
 
 
 class RTHourlyAggregation(PartitionedGCSArtifact):
@@ -277,7 +281,7 @@ class MostRecentSchedule:
                     .get_url_schedule(self.base64_validation_url)
                 )
             except KeyError:
-                print(
+                typer.secho(
                     f"no schedule data found for {self.base64_validation_url} on day {day}"
                 )
                 continue
@@ -287,7 +291,7 @@ class MostRecentSchedule:
                 self.fs.get(schedule_extract.path, gtfs_zip)
                 return gtfs_zip
             except FileNotFoundError:
-                print(
+                typer.secho(
                     f"no schedule file found for {self.base64_validation_url} on day {day}"
                 )
                 continue
@@ -346,17 +350,17 @@ class AggregationExtracts:
     def get_results_paths(self) -> Dict[str, GTFSRTFeedExtract]:
         return {e.get_results_path(): e.extract for e in self.get_extracts()}
 
-    def get_hashed_results(self):
+    def get_hashed_results(self) -> Dict[str, Any]:
         hashed = {}
         for e in self.get_extracts():
             if e.has_results():
-                hashed[e.hash()] = e.get_results()
+                hashed[e.hash().hex()] = e.get_results()
         return hashed
 
-    def get_hashes(self) -> Dict[bytes, List[GTFSRTFeedExtract]]:
-        hashed: Dict[bytes, List[GTFSRTFeedExtract]] = defaultdict(list)
+    def get_hashes(self) -> Dict[str, List[GTFSRTFeedExtract]]:
+        hashed: Dict[str, List[GTFSRTFeedExtract]] = defaultdict(list)
         for e in self.get_extracts():
-            hashed[e.hash()].append(e.extract)
+            hashed[e.hash().hex()].append(e.extract)
         return hashed
 
     def download(self):
@@ -507,7 +511,7 @@ class ValidationProcessor:
             e = ScheduleDataNotFound(
                 f"no recent schedule data found for {self.aggregation.extracts[0].path}"
             )
-            print(e)
+            typer.secho(e)
 
             scope.fingerprint = [
                 type(e),
@@ -571,11 +575,11 @@ class ValidationProcessor:
             for hash, extracts in aggregation_extracts.get_hashes().items():
                 try:
                     records = hashed_results[hash]
-                except KeyError as e:
+                except KeyError:
                     if self.verbose:
                         paths = ", ".join(e.path for e in extracts)
                         typer.secho(
-                            f"WARNING: no results found for {paths}",
+                            f"WARNING: validator did not produce results for {paths}",
                             fg=typer.colors.YELLOW,
                         )
 
@@ -584,7 +588,7 @@ class ValidationProcessor:
                             RTFileProcessingOutcome(
                                 step=self.aggregation.step,
                                 success=False,
-                                exception=e,
+                                exception=NoValidatorResults("No validator output"),
                                 extract=extract,
                             )
                         )
@@ -680,7 +684,7 @@ class ParseProcessor:
                 except DecodeError as e:
                     if self.verbose:
                         typer.secho(
-                            f"WARNING: DecodeError for {str(extract.path)}",
+                            f'DecodeError: "{str(e)}" thrown when decoding {str(extract.path)}',
                             fg=typer.colors.YELLOW,
                         )
                     outcomes.append(
@@ -918,13 +922,9 @@ def main(
         # TODO: I dislike having to exclude the records here
         #   I need to figure out the best way to have a single type represent the "metadata" of
         #   the content as well as the content itself
-        result.save_content(
-            fs=get_fs(),
-            content="\n".join(
-                (json.dumps(make_pydantic_model_bq_safe(o)) for o in result.outcomes)
-            ).encode(),
-            exclude={"outcomes"},
-        )
+        raw = [json.dumps(make_pydantic_model_bq_safe(o)) for o in result.outcomes]
+        content = "\n".join(raw).encode("utf-8")
+        result.save_content(fs=get_fs(), content=content, exclude={"outcomes"})
 
     assert (
         len(outcomes)

--- a/jobs/gtfs-rt-parser-v2/pyproject.toml
+++ b/jobs/gtfs-rt-parser-v2/pyproject.toml
@@ -5,7 +5,7 @@ description = ""
 authors = ["Andrew Vaccaro <atvaccaro@gmail.com>"]
 
 [tool.poetry.dependencies]
-python = ">=3.8,<3.10"
+python = ">=3.8.1,<3.10"
 gtfs-realtime-bindings = "0.0.7"
 google-auth = "1.32.1"
 pathy = {extras = ["gcs"], version = "^0.6.1"}
@@ -26,6 +26,7 @@ types-protobuf = "^5.28.0.20240924"
 types-tqdm = "^4.66.0.20240417"
 isort = "^5.13.2"
 pytest-env = "^1.1.5"
+flake8 = "^7.1.1"
 
 [build-system]
 requires = ["poetry-core>=1.0.0"]

